exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterSink.java

```diff
@@ -202,32 +209,36 @@ public void setJoinMjId2ScanMjId(Map<Integer, Integer> joinMjId2ScanMjId)
     @Override
     public void run()
     {
-      while (joinMjId2rfNumber == null || !joinMjId2rfNumber.isEmpty()) {
-        queueLock.lock();
-        RuntimeFilterWritable toAggregate = rfQueue.poll();
-        if (toAggregate == null) {
-          try {
-            notEmpty.await();
-          }
-          catch (InterruptedException e) {
-            logger.error("RFW_Aggregator thread being interrupted", e);
-          }
-          finally {
-            queueLock.unlock();
-          }
-        } else {
-          queueLock.unlock();
+      while ((joinMjId2rfNumber == null || !joinMjId2rfNumber.isEmpty() ) && running.get()) {
+        RuntimeFilterWritable toAggregate = null;
+        synchronized (rfQueue) {
           try {
-            aggregate(toAggregate);
-          }
-          catch (Exception e) {
-            logger.error("fail to aggregate or route the RFW", e);
-            throw new RuntimeException(e);
-          }
-          finally {
-            toAggregate.close();
+            toAggregate = rfQueue.poll();
+            while (toAggregate == null && running.get()) {
+              rfQueue.wait();
+              toAggregate = rfQueue.poll();
+            }
+          } catch (InterruptedException ex) {
+            logger.error("RFW_Aggregator thread being interrupted", ex);
+            continue;
           }
         }
+        // perform aggregate outside the sync block.
+        try {
+          aggregate(toAggregate);
```

Comment:
check if `toAggregate != null` and then only call `aggregate(toAggregate)`. This can happen for case when `running` is false.